plugin/core/input_handlers.py

```diff
@@ -0,0 +1,205 @@
+from .constants import ST_VERSION
+from .typing import Any, Callable, Dict, List, Optional, ParamSpec, Tuple, Union
+from .typing import final
+from abc import ABCMeta
+from abc import abstractmethod
+import functools
+import sublime
+import sublime_plugin
+import threading
+import weakref
+
+
+ListItemsReturn = Union[List[str], Tuple[List[str], int], List[Tuple[str, Any]], Tuple[List[Tuple[str, Any]], int],
+                        List[sublime.ListInputItem], Tuple[List[sublime.ListInputItem], int]]
+
+P = ParamSpec('P')
+
+
+def debounced(user_function: Callable[P, Any]) -> Callable[P, None]:
+    """ A decorator which debounces the calls to a function.
+
+    Note that the return value of the function will be discarded, so it only makes sense to use this decorator for
+    functions that return None.
+    """
+    DEBOUNCE_TIME = 0.5  # seconds
+
+    @functools.wraps(user_function)
+    def wrapped_function(*args: P.args, **kwargs: P.kwargs) -> None:
+        def call_function():
+            if hasattr(wrapped_function, '_timer'):
+                delattr(wrapped_function, '_timer')
+            return user_function(*args, **kwargs)
+        timer = getattr(wrapped_function, '_timer', None)
+        if timer is not None:
+            timer.cancel()
+        timer = threading.Timer(DEBOUNCE_TIME, call_function)
+        timer.start()
```

Comment:
This runs user function on a separate thread. Can create synchronization/timing issues if user function relies on data set on main (or sublime's async) thread.

You could use set_timeout instead. Even if it would require some less than ideal tracking of monotonic IDs or similar.

Reply:
I think this should be fixed now and the user function runs on the UI thread.
But it would be good if you could take a second look to confirm.

I'm not sure whether the `WorkspaceSymbolsInputHandler.on_modified` method is now blocking the UI thread due to the `Promise.all(...)`. On a quick test with the workspace symbols it still seems to work fine and I couldn't see any lag while typing. But that doesn't have to mean much.

Reply:
Perhaps it should always use the async thread and then we rename the method in `DynamicListInputHandler` to `on_modified_async`?

Reply:
I was thinking more of getting rid of creating new threads completely because if the intention of such debouncing decorator is to debounce frequently running function then the overhead of creating new thread on creating each new timer feels quite high.

Of course with set timeout you can't cancel so would need to track and ignore obsolete debounces.

As for async vs sync, maybe it makes sense to add an option for that already but I wouldn't mind doing that when needed later

Reply:
I guess the new thread is only created when the timeout actually "timeouts". So it shouldn't be as bad as I thought. But still there is at least one new thread created which seems wasteful because we redirect to the main thread anyway

Reply:
Actually I believe it creates a new thread each time when the debounced function is called, because `threading.Timer` immediately creates a new thread. There will never be multiple threads running in parallel, because the last one is cancelled before a new thread gets created, but it might still be a little bit of overhead from the new thread creation. But I don't know how much this overhead is and how it compares to the set_timeout API calls, which I guess also require some kind of scheduling logic with timers implemented in ST.

There was a nice little trick recently posted in the forum where you would store the current timestamp and then instead of queuing a new callback with set_timeout (or correspondigly restarting the thread in the implementation from this PR) it would just update the target time when the function should be called. Then, when the initial timeout has elapsed, it will first compare the current time with the target time and add another delay with set_timeout if there is a difference. This is certainly useful for the set_timeout solution, because there is no way to cancel those.

I could probably do the same here, but I don't know how the performance of the current thread restarting compares to the `time.time()` calls and the additional logic that would be needed. So maybe it would be better to postpone such performance testing and optimizations until someone observes any performance issues in practice with the current implementation. Btw, the same could also be applied to the existing debounce implementations in
https://github.com/sublimelsp/LSP/blob/df4b5e44916f3afc860204f257c3ca40a92abb96/plugin/core/types.py#L116
and
https://github.com/sublimelsp/LSP/blob/df4b5e44916f3afc860204f257c3ca40a92abb96/plugin/core/types.py#L147

Reply:
set_timeout doesn't create new thread. It can be verified with process manager on macos at least.

I would prefer to avoid using `threading.Timer` already since it's likely to be used more in the future and then it might be overlooked how it's implemented.

Reply:
> There was a nice little trick recently posted in the forum where you would store the current timestamp and then instead of queuing a new callback with set_timeout (or correspondigly restarting the thread in the implementation from this PR) it would just update the target time when the function should be called. Then, when the initial timeout has elapsed, it will first compare the current time with the target time and add another delay with set_timeout if there is a difference. This is certainly useful for the set_timeout solution, because there is no way to cancel those.

Is it maybe a bit unnecessarily complicated? In other places we just have a variable that tracks the ID bound to the latest `set_timeout` invocation and we ignore timeouts with non-latest ID. If the ID is scoped to the debounced function then it should work in a simple and effective way?

Reply:
I think the approach with just queuing new `set_timeout`s with checking an ID isn't better than the timer restarting that is currently implemented here. But I can look to implement the smarter logic with tracking time differences and then use `set_timeout`. I think the way it was posted in the forum has some room for further optimizations though, because it assumes that the delay from `set_timeout` is always exact. If there would be a small bias for the sleeping time introduced in the future, it might lead to an "almost infinite" loop. But this should be possible to avoid by checking an update variable/ID instead of the time difference.

---

Btw the approach with `threading.Timer` is heavily inspired from https://github.com/TerminalFi/LSP-copilot/blob/1fe28f139031e1756ffd8da5c7d025872233344d/plugin/utils.py#L54, and they seem to use it for a function called in `on_modified_async` (i.e. much more often than here), and apparently without complains.

Reply:
I pushed a version that doesn't spawn the Timer thread.

Reply:
Sorry, missed the update.